File: task4.cpp
#include <iostream>
#include <vector>
#include <random>
#include <locale.h>
using namespace std;

// Алгоритм "Доверчивый"
bool confiding(int round_number, vector<bool> self_choices, vector<bool> enemy_choices) 
{
    // Всегда сотрудничает, независимо от предыдущих ходов
    return true;
}

// Алгоритм "Мстительный"
bool revengeful(int round_number, vector<bool> self_choices, vector<bool> enemy_choices) 
{
    // Сотрудничает, если в предыдущем раунде противник также сотрудничал, иначе предает
    if (round_number == 0 || enemy_choices[round_number - 1]) 
    {
        return true;
    }
    else {
        return false;
    }
}

// Алгоритм "Осторожный"
bool careful(int round_number, vector<bool> self_choices, vector<bool> enemy_choices) 
{
    // Сотрудничает, если в предыдущих 3 раундах противник 2 раза или больше сотрудничал, иначе предает
    int cooperationCount = 0;
    for (int i = max(0, round_number - 3); i < round_number; i++) 
    {
        if (enemy_choices[i]) 
        {
            cooperationCount++;
        }
    }
    return cooperationCount >= 2;
}

int main() 
{
    setlocale(LC_ALL, "Rus");
    // Генерируем случайное количество раундов от 100 до 200
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distribution(100, 200);
    int total_rounds = distribution(gen);

    // Массивы для хранения выборов игроков
    vector<bool> player1_choices(total_rounds, false);
    vector<bool> player2_choices(total_rounds, false);

    // Выбор алгоритмов для игроков
    bool (*player1_algorithm)(int, vector<bool>, vector<bool>) = &revengeful;
    bool (*player2_algorithm)(int, vector<bool>, vector<bool>) = &confiding;

    // Игровой цикл
    int player1_score = 0, player2_score = 0;
    for (int round = 0; round < total_rounds; round++) {
        player1_choices[round] = (*player1_algorithm)(round, player1_choices, player2_choices);
        player2_choices[round] = (*player2_algorithm)(round, player2_choices, player1_choices);

        if (player1_choices[round] && player2_choices[round]) 
        {
            player1_score += 24;
            player2_score += 24;
        }
        else if (player1_choices[round] && !player2_choices[round]) 
        {
            player2_score += 20;
        }
        else if (!player1_choices[round] && player2_choices[round]) 
        {
            player1_score += 20;
        }
        else 
        {
            player1_score += 4;
            player2_score += 4;
        }
    }

    // Вывод результатов
    cout << "Количество раундов: " << total_rounds << endl;
    cout << "Счет игрока 1: " << player1_score << endl;
    cout << "Счет игрока 2: " << player2_score << endl;

    return 0;
}

File: Go/task4.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// Алгоритм "Доверчивый"
func confiding(roundNumber int, selfChoices, enemyChoices []bool) bool {
	// Всегда сотрудничает, независимо от предыдущих ходов
	return true
}

// Алгоритм "Мстительный"
func revengeful(roundNumber int, selfChoices, enemyChoices []bool) bool {
	// Сотрудничает, если в предыдущем раунде противник также сотрудничал, иначе предает
	if roundNumber == 0 || enemyChoices[roundNumber-1] {
		return true
	}
	return false
}

// Алгоритм "Осторожный"
func careful(roundNumber int, selfChoices, enemyChoices []bool) bool {
	// Сотрудничает, если в предыдущих 3 раундах противник 2 раза или больше сотрудничал, иначе предает
	cooperationCount := 0
	for i := max(0, roundNumber-3); i < roundNumber; i++ {
		if enemyChoices[i] {
			cooperationCount++
		}
	}
	return cooperationCount >= 2
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	// Генерируем случайное количество раундов от 100 до 200
	rand.Seed(time.Now().UnixNano())
	totalRounds := rand.Intn(101) + 100

	// Массивы для хранения выборов игроков
	player1Choices := make([]bool, totalRounds)
	player2Choices := make([]bool, totalRounds)

	// Выбор алгоритмов для игроков
	player1Algorithm := revengeful
	player2Algorithm := confiding

	// Игровой цикл
	player1Score, player2Score := 0, 0
	for round := 0; round < totalRounds; round++ {
		player1Choices[round] = player1Algorithm(round, player1Choices, player2Choices)
		player2Choices[round] = player2Algorithm(round, player2Choices, player1Choices)

		if player1Choices[round] && player2Choices[round] {
			player1Score += 24
			player2Score += 24
		} else if player1Choices[round] && !player2Choices[round] {
			player2Score += 20
		} else if !player1Choices[round] && player2Choices[round] {
			player1Score += 20
		} else {
			player1Score += 4
			player2Score += 4
		}
	}

	// Вывод результатов
	fmt.Println("Количество раундов:", totalRounds)
	fmt.Println("Счет игрока 1:", player1Score)
	fmt.Println("Счет игрока 2:", player2Score)
}
